boards/p1am_100/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+    }
+    PA17 {
+        name: d9,
+        aliases: {
+            AlternateC: Spi0Sck
+        }
+    }
+    PA18 {
+        name: pa18,
+        aliases: {
+            /// Host Enable, drive high to switch into USB host mode
+            #[cfg(feature = "usb")]
+            PushPullOutput: UsbId
+        }
+    }
+    PA19 {
+        name: d10,
+        aliases: {
+            AlternateC: Spi0Miso
+        }
+    }
+    PA20 { name: d6 }
+    PA21 { name: d7 }
+    PA22 { name: d0 }
+    PA23 { name: d1 }
+    PA24 {
+        name: usb_dm,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDm
+        }
+    }
+    PA25 {
+        name: usb_dp,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDp
+        }
+    }
+    PA27 {
+        name: pa27,
+        aliases: {
+            PullUpInput: SdCardDetect
+        }
+    }
+    PA28 {
+        name: switch,
+        aliases: {
+            PullUpInput: Switch
+        }
+    }
+    PB02 {
+        name: d16,
+        aliases: {
+            AlternateB: A1
+        }
+    }
+    PB03 {
+        name: d17,
+        aliases: {
+            AlternateB: A2
+        }
+    }
+    PB08 {
+        name: led,
+        aliases: {
+            PushPullOutput: Led
+        }
+    }
+    PB09 {
+        name: pb09,
+        aliases: {
+            PushPullOutput: BaseEnable,
+            AlternateB: AdcBattery
+        }
+    }
+    PB10 { name: d4 }
+    PB11 { name: d5 }
+    PB22 {
+        name: d14,
+        aliases: {
+            AlternateD: UartTx
+        }
+    }
+    PB23 {
+        name: d13,
+        aliases: {
+            AlternateD: UartRx
+        }
+    }
+);
+
+const BASE_CONTROLLER_FREQ: Hertz = Hertz(1000000);
+// FIXME: const BASE_CONTROLLER_SPI_MODE: dyn spi::Mode = &spi::MODE_2;
+
+/// Convenience for setting up the labeled SPI0 peripheral.
+/// SPI0 has the P1AM base controller connected.
+/// This powers up SERCOM1 and configures it for talking to the
+/// base controller.
+pub fn base_controller_spi(
+    clocks: &mut GenericClockController,
+    sercom1: pac::SERCOM1,
+    pm: &mut pac::PM,
+    sck: Spi0Sck,
+    mosi: Spi0Mosi,
+    miso: Spi0Miso,
+) -> sercom::v2::spi::Spi<impl spi::ValidConfig> {
+    let gclk0 = &clocks.gclk0();
+    let core_clock = &clocks.sercom1_core(&gclk0).unwrap();
+    let pads = spi::Pads::new().sclk(sck).data_in(miso).data_out(mosi);
+    spi::Config::new(pm, sercom1, pads, core_clock.freq())
+        .baud(BASE_CONTROLLER_FREQ)
+        .spi_mode(spi::MODE_2)
+        .enable()
+}
+
+/// Convenience for setting up the labelled SDA, SCL pins to
+/// operate as an I2C master running at the specified frequency.
+pub fn i2c_master<F: Into<Hertz>>(
+    clocks: &mut GenericClockController,
+    bus_speed: F,
+    sercom0: pac::SERCOM0,
+    pm: &mut pac::PM,
+    sda: Sda,
+    scl: Scl,
+    port: &mut Port,
+) -> I2CMaster0<
+    impl hal::sercom::AnyPad<Sercom = pac::SERCOM0>,
```

Comment:
Again, I think I understand why you did it this way. The difference between the `v1` and `v2` `Pad` types was confusing, and the `Map` trait was overly complex. In the latest iteration, the two versions are totally separate types.

If you want to update to `v2` `Pad` types right now, you could use

```rust
use hal::sercom::v2::{Pad, Sercom0, Pad0, Pad1};
// ...
I2CMaster0<
    Pad<Sercom0, Pad0, <Sda as AnyPin>::Id>,
    Pad<Sercom0, Pad0, <Scl as AnyPin>::Id>,
>
```

In the latest iteration, I also added a type alias `SpecificPinId<P>` to use instead of the `<P as AnyPin>::Id` syntax.

Reply:
I don't want to have to write the pad number in this type. I want the compiler to infer it from the sercom and the pin.

Reply:
I'm still not sure I quite understand the motivation for inferring types in situations like this. The output type is directly linked to the `Scl`, etc. input types. There would be no way to change the output type without also changing the input types, so all changes are breaking. And by using existential types, you only serve to hide potentially useful information from the caller.

If that's how you want to handle it, then feel free. I'm just trying to point out that there may be potentially unintended consequences when you hide type information (see the comment on `Spi` above). The HAL uses types to ensure valid configurations, so hiding types can lead to problems.

Reply:
FYI, I also plan to add `PinId` aliases to the `bsp_pins!` macro. So you could use something like `SclId` in place of `SpecificPinId<Scl>` or `<Scl as AnyPin>::Id`.